src/main/java/hudson/remoting/RemoteClassLoader.java

```diff
@@ -330,8 +332,17 @@ private void invokeClassLoadTestingHookIfNeeded() throws InterruptedException {
         }
     }
 
-    private boolean shouldRetry(Throwable e) {
-        return e instanceof InterruptedException || (e instanceof RemotingSystemException
+    private boolean shouldRetry(Throwable e, int tries) {
+        return isRetryException(e) && hasMoreRetries(tries);
+    }
+
+    private boolean hasMoreRetries(int tries) {
+        return MAX_RETRIES <= 0 || tries <= MAX_RETRIES;
```

Comment:
I was more expecting `MAX_RETRIES=0` as a way to disable the retry mechanism.

Maybe something like this? `return MAX_RETRIES < 0 || tries <= MAX_RETRIES;`
- positive number: max number of retries
- zero: disable retries
- negative: infinite retries